### PR TITLE
New metric holder

### LINKED ISSUE #410: Feature: New Metric Holder
Need new metric holder with next features:

- No dependency on action-plugins metrics
- Easy expandable
- Once in a certain period of time, all metrics that havn't been used since the last update should be deleted

### PR DESCRIPTION
**Pull request by kirillov6**
# Description

Features:
- No dependency on action-plugins metrics
- Easy expandable
- Once in a certain period of time, all metrics that haven't been used since the last update will be deleted

Fixes #410

### CODE REVIEW

File: metric/holder.go [hunk 1]
```diff
@@ -0,0 +1,271 @@
+package metric
+
+import (
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const genTimeInterval = time.Second
```
Comment (vadimalekseev) [inline metric/holder.go:10]: did you mean setTimeInterval?

File: metric/holder.go [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+
+var (
+	nowTime   time.Time
```
Comment (vadimalekseev) [inline metric/holder.go:13]: Can we use atomic.Int64 here?

```go
var nowTime = atomic.Int64{}

func setNowTime(t time.Time) {
  nowTime = t.UnixNano()
}

func getNowTime() time.Time {
  return time.Unix(0, nowTime.Load())
}
```

Comment (vadimalekseev) [reply metric/holder.go:13]: I think we can move this logic to a new package `github.com/ozontech/file.d/xtime.GetInaccurateTime()`

File: metric/controller.go [hunk 3]
```diff
@@ -19,90 +19,145 @@ type Ctl struct {
 	subsystem string
 	register  *prometheus.Registry
 
-	counters  map[string]*prometheus.CounterVec
-	counterMx *sync.Mutex
+	counters  map[string]prometheus.Counter
+	counterMx *sync.RWMutex
```
[the remark this answers is not in the document]
Comment (kirillov6) [reply metric/controller.go]: Done

File: metric/holder.go (hunk 1 above)
Comment (kirillov6) [reply metric/holder.go]: Now it's `updateTimeInterval` in xtime package

File: xtime/xtime.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import (
+	"sync/atomic"
+	"time"
+)
+
+const updateTimeInterval = time.Second
+
+var nowTime atomic.Int64
+
+func GetInaccurateTimeUnix() int64 {
```
Comment (vadimalekseev) [inline xtime/xtime.go:12]:
```suggestion
func GetInaccurateTimeUnixNano() int64 {
```

File: xtime/xtime.go [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+	return nowTime.Load()
+}
+
+func GetInaccurateTime() time.Time {
+	return time.Unix(0, nowTime.Load())
+}
+
+func Start() {
```
Comment (vadimalekseev) [inline xtime/xtime.go:20]:
```suggestion
var startOnce = sync.OnceFunc(func() {
  SetNowTime(time.Now())
  ticker := time.NewTicker(updateTimeInterval)
  for t := range ticker.C {
    SetNowTime(t)
  }
})

func GetInaccurateTimeUnixNano() int64 {
  startOnce()
  return nowTime.Load()
}

func GetInaccurateTime() time.Time {
  time.Unix(0, GetInaccurateTimeUnixNano())
}
```

File: xtime/xtime.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+	ticker := time.NewTicker(updateTimeInterval)
+	for t := range ticker.C {
+		SetNowTime(t)
+	}
+}
+
+func SetNowTime(t time.Time) {
```
Comment (vadimalekseev) [inline xtime/xtime.go:27]: I think it must be private

File: xtime/xtime.go (hunk 4 above)
Comment (kirillov6) [reply xtime/xtime.go]: Done

File: xtime/xtime.go (hunk 5 above)
Comment (kirillov6) [reply xtime/xtime.go]: Done

File: xtime/xtime.go (hunk 6 above)
Comment (kirillov6) [reply xtime/xtime.go]: Done

**Review by vadimalekseev — APPROVED** [review]

**Review by vano144 — APPROVED** [review]

**Pull request merged**